File: src/main/java/com/dunapart/ParkoloApp/Backend/DataController.java
package com.dunapart.ParkoloApp.Backend;

import com.dunapart.ParkoloApp.APISpringManager;
import com.dunapart.ParkoloApp.Backend.Model.LoginRequest;
import com.dunapart.ParkoloApp.Backend.Model.RegiRequest;
import com.dunapart.ParkoloApp.JwtUtil;
import org.hibernate.Hibernate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@RestController
@RequestMapping("/api")
@SpringBootApplication
public class DataController {

    private final APISpringManager springmanager;
    private final JwtUtil jwtUtil; // JWT segédosztály


    @Autowired
    public DataController(APISpringManager springmanager) {
        this.springmanager = springmanager;
        this.jwtUtil = new JwtUtil();

    }

    @PostMapping("/login")
    public ResponseEntity<?> getFelhasznalo(@RequestBody LoginRequest loginRequest) {

        System.out.println("email: " + loginRequest.getEmail());
        System.out.println("passwd: " + loginRequest.getPassword());


        String validity = springmanager.isUserValid(loginRequest.getEmail(), loginRequest.getPassword());

        if ("valid".equals(validity)) {
            String token = jwtUtil.generateToken(loginRequest.getEmail()); // Token generálás
            return ResponseEntity.ok(Map.of("token", token)); // Token visszaküldése
        } else {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid login credentials!");
        }
    }

    @PostMapping("/reg")
    public ResponseEntity<?> regFelhasznalo(@RequestBody RegiRequest regiRequest)
    {
        System.out.println("regelt felhasznalo adatai: ");
        System.out.println("firstname: "+regiRequest.getFirstName());
        System.out.println("lastname: "+regiRequest.getLastName());
        System.out.println("passwd: "+regiRequest.getPassword());
        System.out.println("email: "+regiRequest.getEmail());
        springmanager.saveUser(regiRequest.getFirstName(),regiRequest.getLastName(),regiRequest.getPassword(),regiRequest.getEmail());
        System.out.println("user reg method finished");

        Map<String, String> response = new HashMap<>();
        response.put("message", "User successfully registered");
        response.put("status", "success");
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @GetMapping("/me")
    public ResponseEntity<?> getUserData(@RequestHeader("Authorization") String authHeader) {
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Missing or invalid token");
        }

        String token = authHeader.substring(7); // "Bearer " eltávolítása
        if (!jwtUtil.isTokenValid(token)) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid token");
        }

        String email = jwtUtil.extractEmail(token); // Tokenből e-mail kinyerése
        // Példa: lekérdezzük a felhasználót a SpringManager-ből
        Felhasznalo user = springmanager.findUserByEmail(email);
        return ResponseEntity.ok(user);
    }

    @GetMapping("/loadCars")
    public ResponseEntity<?> getUserCars(@RequestHeader("email") String userHeader) {
        if (userHeader == null || !userHeader.startsWith("Bearer ")) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Missing or invalid email");
        }

        try
        {
            String email = userHeader.substring(7); // email kinyerése
            Felhasznalo user = springmanager.findUserByEmail(email);
            long userID = user.getFelhasznalo_id();

            List<Autok> userCars = springmanager.findCars(userID);
//            userCars.forEach(car -> Hibernate.initialize(car.getParkolo()));
            return ResponseEntity.ok(userCars);
        }
        catch(Exception ex)
        {
            System.out.println("A hiba:::::::" + ex.toString() + "\n" + ex.getMessage());
            return null;
        }

    }


    @PostMapping("/addCar")
    public ResponseEntity<?> addUserCar(@RequestHeader("email") String email, @RequestHeader("meret") int meret, @RequestHeader("rendszam") String rendszam,
                                        @RequestHeader("color") String color, @RequestHeader("name") String name,@RequestHeader("type") String type)
    {
        Felhasznalo felhasznalo = springmanager.findUserByEmail(email);

        if(springmanager.isRendszamExist(rendszam))
        {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Ilyen rendszámmal már létezik autó!");
        }

        String itWasASuccess = springmanager.addUserCar(felhasznalo, meret, rendszam, color,name,type);
        if(itWasASuccess.equals("OK"))
        {
            return ResponseEntity.status(HttpStatus.CREATED).body("Az autó hozzáadásra került");
        }
        else
        {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Az autó nem került hozzáadásra");
        }
    }
    @PostMapping("/editCar")
    public ResponseEntity<?> editUserCar(@RequestHeader("email") String userHeader, @RequestHeader("rendszam") String rendszam,
                                         @RequestHeader("color") String color, @RequestHeader("name") String name,@RequestHeader("type") String type, @RequestHeader("auto_id") int ID) {


        if (rendszam == null) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Missing or invalid rendszam");
        }

        String email = userHeader;
        Felhasznalo user = springmanager.findUserByEmail(email);
        long userID = user.getFelhasznalo_id();
        List<Autok> userCars = springmanager.findCars(userID);
        Autok userCar = springmanager.findCarByID(userCars, ID);

        System.out.println(userCar);

        String succes = springmanager.updateCarByID(userCar, rendszam,color,name,type);
        if(succes.equals("OK"))
        {
            return ResponseEntity.status(HttpStatus.OK).body(userCar.getAuto_id() + " Frissitesre sikeres");
        }
        else
        {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Az auto adatainak frissitese sikertelen!");
        }
    }

    @PostMapping("/deleteCar")
    public ResponseEntity<?> deleteUserCar(@RequestHeader("email") String userHeader,@RequestHeader("auto_id") long ID) {

        if(ID == 0)
        {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Missing or invalid ID");
        }
        else
        {
            String email = userHeader;
            Felhasznalo user = springmanager.findUserByEmail(email);
            long userID = user.getFelhasznalo_id();
            List<Autok> userCars = springmanager.findCars(userID);
            Autok userCar = springmanager.findCarByID(userCars, ID);

            String succes = springmanager.deleteCarByID(userCar);
            if(succes.equals("OK"))
            {
                return ResponseEntity.status(HttpStatus.OK).body("Autó törlés sikeres");
            }
            else if(succes.equals("nOK"))
            {
                return ResponseEntity.status(HttpStatus.CONFLICT).body("Sikertelen autó törlés");
            }
            else
            {
                return ResponseEntity.status(HttpStatus.CONFLICT).body("Sikertelen autó törlés");
            }
        }
    }

    @PostMapping("/forgotPassword")
    public ResponseEntity<?> resetPassword( @RequestHeader("email") String email,@RequestHeader("newpasswd") String newpasswd)
    {
        try
        {
            Felhasznalo user = springmanager.findUserByEmail(email);
            springmanager.saveUser(user.getFirstName(),user.getLastName(),newpasswd,user.getEmail());
            return ResponseEntity.status(HttpStatus.OK).body("Sikeres jelszó módosítás");
        }
        catch(Exception ex)
        {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Sikertelen jelszó módosítás");
        }
    }

    @GetMapping("/loadUserPlot")
    public ResponseEntity<?> getUserPlots(@RequestHeader("email") String userHeader) {
        if (userHeader == null || !userHeader.startsWith("Bearer ")) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("Missing or invalid email");
        }

        String email = userHeader.substring(7);
        Felhasznalo user = springmanager.findUserByEmail(email);
        long userID = user.getFelhasznalo_id();
        List<Autok> userCars = new ArrayList<>();
        userCars = springmanager.findCars(userID);

        List<Long> userParkolo = new ArrayList<>();

        for (Autok car : userCars) {
            userParkolo.add(car.getParkolo().getParkolo_id());
        }

        return ResponseEntity.ok(userParkolo);
    }

    @GetMapping("/loadPlots")
    public ResponseEntity<?> getFreePlots() {
        List<Parkolo> parkolohelyek = springmanager.getParkingPlots(); //itt lekérem az összeset
        return ResponseEntity.ok(parkolohelyek);
    }

    //frontendről meg kell kapni azt is hogy melyik autót akarja parkoltatni a felhasználó
    //az auto_id + parkolo_id-t megkapja a frontendről
    @PostMapping("/saveBookingDate")
    public ResponseEntity<?> saveBookingDate(@RequestHeader("to_date") String to_date_string, @RequestHeader("from_date") String from_date_string, @RequestHeader("auto_id") long auto_id, @RequestHeader("parkolo_id") long selected_plot)
    {
        //time stuffs
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime to_date = LocalDateTime.parse(to_date_string,formatter);
        LocalDateTime from_date = LocalDateTime.parse(from_date_string,formatter);

        //repo stuffs
        Autok wanna_save_auto = springmanager.getUserCarById(auto_id);
        Parkolo wanna_park_here = springmanager.getParkingPlotById(selected_plot);

        //vizsgálat ha a parkoló mérete kisebb mint az autó mérete akkor hibát neki had rágja
        if(wanna_save_auto.getMeret() < wanna_park_here.getMeret())
        {
            return ResponseEntity.status(HttpStatus.PRECONDITION_FAILED).body("Hiba! Nem megfelelő parkolóméret");
        }

        String parkoloFrissitese = springmanager.updateParkoloById(wanna_park_here, to_date, from_date);
        String autoFrissitese = springmanager.updateAuto(wanna_save_auto,wanna_park_here);

        if(parkoloFrissitese.equals("OK") && autoFrissitese.equals("OK"))//frissítésre került
        {
            return ResponseEntity.status(HttpStatus.CREATED).body("Foglalás hozzáadva");
        }
        else
        {
            return ResponseEntity.status(HttpStatus.EXPECTATION_FAILED).body("Hiba a foglalás mentése során");
        }

    }

    @PostMapping("/deleteBookedPlot")
    public ResponseEntity<?> deleteBookedPlot(@RequestHeader("parkolo_id") long parkolo_id, @RequestHeader("auto_id") long auto_id)
    {
        String result = springmanager.deleteBooking(parkolo_id, auto_id);
        if(result.equals("OK"))
        {
            return ResponseEntity.status(HttpStatus.OK).body("Foglalás törölve");
        }
        else
        {
            return ResponseEntity.status(HttpStatus.EXPECTATION_FAILED).body("Foglalás nem lett törölve");
        }

    }

    @PostMapping("/saveBookingDate")
    public ResponseEntity<?> saveBookingDate(@RequestHeader("to_date") String to_date_string, @RequestHeader("parkolo_id") long selected_plot)
    {
        //time stuffs
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
        LocalDateTime to_date = LocalDateTime.parse(to_date_string,formatter);

        String result = springmanager.modifyBookingDate(to_date, selected_plot);
        if(result.equals("OK"))
        {
            return ResponseEntity.status(HttpStatus.OK).body("A foglalás meghosszabbítva");
        }
        else
        {
            return ResponseEntity.status(HttpStatus.EXPECTATION_FAILED).body("Foglalás nem lett meghosszabbítva");
        }
    }

}

File: src/api/dataController.js
// api.js
// api.js

export const fetchUserData = async (token) => {
  const response = await fetch('http://localhost:8084/api/me', {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`, // Token használata autentikációhoz
    },
  });

  if (!response.ok) {
    throw new Error('Nem sikerült lekérni a felhasználói adatokat.');
  }
  
  const userData = await response.json(); // Válasz JSON formátumra alakítása
  return userData;
};

export const fetchCarsData = async (userEmail) => {
  const response = await fetch('http://localhost:8084/api/loadCars', {
    method: 'GET',
    headers: {
      email: `Bearer ${userEmail}`, // Email használata az autók lekéréséhez
    },
  });

  if (!response.ok) {
    throw new Error('Nem sikerült betölteni az autók listáját.');
  }
  
  const carData = await response.json(); // Válasz JSON formátumra alakítása
  return carData;
};

export const addCar = async (carData) => {
  const { email, meret, rendszam, color, name, type } = carData; // Az autó adatai objektumként kerülnek átadásra
  const response = await fetch('http://localhost:8084/api/addCar', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      email, // Felhasználói email
      meret,
      rendszam, // Rendszám
      color, // Szín
      name, // Név
      type, // Típus
    },
  });

  if (!response.ok) {
    if (response.status === 409) {
      throw new Error('Ilyen rendszámmal már létezik autó!');
    }
    throw new Error('Az autó nem került hozzáadásra.');
  }

  const result = await response.text(); // Backend válasza
  return result;
};

export const editCar = async (carData) => {
  const { email, auto_id, rendszam, color, name, type } = carData; // Az autó adatai objektumként kerülnek átadásra
  const response = await fetch('http://localhost:8084/api/editCar', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      email, // Felhasználói email
      rendszam, // Új rendszám
      color, // Új szín
      name, // Új név
      type, // Új típus
      auto_id, // Azonosító az autóhoz
    },
  });

  if (!response.ok) {
    if (response.status === 400) {
      throw new Error('Hiányzó vagy érvénytelen rendszám.');
    }
    if (response.status === 409) {
      throw new Error('Az autó adatainak frissítése sikertelen.');
    }
    throw new Error('Hiba történt az autó frissítése során.');
  }

  const result = await response.text(); // Backend válasza
  return result;
};

export const deleteCar = async (carData) => {
  const { email, auto_id } = carData; // Az autó adatainak szükséges elemei

  const response = await fetch('http://localhost:8084/api/deleteCar', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      email, // Felhasználói email
      auto_id, // Az autó azonosítója
    },
  });
  if (!response.ok) {
    if (response.status === 400) {
      throw new Error('Hiányzó vagy érvénytelen autóazonosító.');
    }
    if (response.status === 409) {
      throw new Error('Az autó törlését nem sikerült végrehajtani.');
    }
    throw new Error('Hiba történt az autó törlése során.');
  }

  const result = await response.text(); // Backend válasza
  return result;
};

export const fetchParkingPlots = async () => {
  try {
    const response = await fetch('http://localhost:8084/api/loadPlots', {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json', // Alapértelmezett content type
      },
    });

    if (!response.ok) {
      throw new Error('Nem sikerült lekérni a parkolóhelyeket.');
    }

    const plotsData = await response.json(); // A válasz JSON formátumra alakítása
    return plotsData; // Visszaadjuk a lekért adatokat
  } catch (error) {
    console.error('Hiba a parkolóhelyek lekérése során:', error);
    throw error; // Hiba dobása, ha a kérés nem sikerült
  }
};

export const saveBookingDate = async (bookingData) => {
  const { to_date, from_date, auto_id, parkolo_id } = bookingData; // Az időpontok és azonosítók objektumként kerülnek átadásra
  const response = await fetch('http://localhost:8084/api/saveBookingDate', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'to_date': to_date, // Formázott dátum (yyyy-MM-dd HH:mm)
      'from_date': from_date, // Formázott dátum (yyyy-MM-dd HH:mm)
      'auto_id': auto_id, // Autó azonosítója
      'parkolo_id': parkolo_id, // Parkoló azonosítója
    },
  });

  if (!response.ok) {
    if (response.status === 412) {
      throw new Error('Nem megfelelő parkolóméret.');
    }
    if (response.status === 417) {
      throw new Error('Hiba a foglalás mentése során.');
    }
    if (response.status === 410) {
      throw new Error('Már parkolsz valahol, barátom.');
    }
    throw new Error('Hiba történt a foglalás mentése során.');
  }

  const result = await response.text(); // Backend válasza
  return result;
};

export const deleteBookedPlot = async (parkolo_id, auto_id) => {
  try {
    const response = await fetch('http://localhost:8084/api/deleteBookedPlot', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'parkolo_id': parkolo_id,
        'auto_id': auto_id,
      },
    });

    if (!response.ok) {
      if (response.status === 412) {
        throw new Error('Nem megfelelő parkolóméret.');
      }
      if (response.status === 417) {
        throw new Error('Hiba a foglalás törlése során.');
      }
      throw new Error('Hiba történt a foglalás törlése során.');
    }

    const result = await response.text(); // Backend válasza
    return result;
  } catch (error) {
    console.error('Hálózati hiba:', error);
    throw error;
  }
};

export const modifyBookingDate = async (to_date, selected_plot) => {
  try {
    const response = await fetch('http://localhost:8084/api/modifyBookingDate', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'to_date': to_date, // Formázott dátum (yyyy-MM-dd HH:mm)
        'parkolo_id': selected_plot, // Parkoló azonosítója
      },
    });

    if (!response.ok) {
      if (response.status === 412) {
        throw new Error('A foglalás meghosszabbítása nem lehetséges.');
      }
      if (response.status === 417) {
        throw new Error('Hiba történt a foglalás módosítása során.');
      }
      throw new Error('Hiba történt a foglalás módosítása során.');
    }

    const result = await response.text(); // A backend válasza
    return result;
  } catch (error) {
    console.error('Hálózati hiba:', error);
    throw error;
  }
};

// resetPassword.js or resetPassword.ts (depending on your setup)
export const resetPassword = async (newPassword, email) => {
  const url = "http://localhost:8084/api/forgotPassword"; // Cseréld le a megfelelő API végpontra

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json", // Ha JSON-t is küldenél, ez szükséges
        "newpasswd": newPassword,
        "email": email
      }
    });

    if (response.ok) {
      const message = await response.text();
      console.log("Sikeres jelszó módosítás:", message);
    } else {
      const error = await response.text();
      console.error("Sikertelen jelszó módosítás:", error);
    }
  } catch (error) {
    console.error("Hiba történt a kérés során:", error);
  }
};

export const clearDatabase = async () => {
  try {
    const response = await fetch('http://localhost:8084/api/clearDatabase', {
      method: 'POST', // POST kérést küldünk
      headers: {
        'Content-Type': 'application/json', // Az adatok formátuma JSON
        // Ha szükséges token vagy egyéb autentikációs fejlécek:
        // 'Authorization': `Bearer ${yourAuthToken}`,
      },
    });

    // Ellenőrizzük, hogy a válasz rendben van-e
    if (response.ok) {
      const data = await response.text(); // Az adatbázis törlés sikerességének üzenete
      console.log(data);
    } else {
      const errorData = await response.text();
      alert(`Hiba történt: ${errorData}`);
    }
  } catch (error) {
    // Hiba kezelése, ha a kérelem nem sikerül
    console.error('Hiba a kérés során:', error);
    alert('Hiba a kérés során');
  }
};
